subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/TestFrameworkAutoDetection.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.testing;
+
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.tasks.testing.Test;
+
+import java.io.File;
+import java.util.LinkedHashSet;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class TestFrameworkAutoDetection {
```

Comment:
I have a longer comment below, but I took a stab at rewriting this to be based on the _direct_ dependencies of the tests:

https://github.com/gradle/gradle/commit/86a1e971699d20d4855c626e9746cf508fc2716e#diff-44b9afd6ec27aaa2a010c1ac97c75191

I think there should be a simple rule (e.g., we prefer this over that), even if sometimes we don't suss out exactly what the user intended.   I think I don't quite have it right because it looks like junit-vintage-engine should select junit platform sometimes.

Maybe we'd also want to make spockframework auto-detect as JUnit.

Reply:
> I think there should be a simple rule (e.g., we prefer this over that), even if sometimes we don't suss out exactly what the user intended. I think I don't quite have it right because it looks like junit-vintage-engine should select junit platform sometimes.

I tried a simple rule initially but it turned to be more complicated (see discussion above wrt. groovy-all). You shouldn't check for `org.junit.jupiter:junit-jupiter` because Jupiter != Platform and that wouldn't work if someone used a different engine (there are actually quite a few by now). Instead, the check should be for `org.junit.platform:junit-platform-engine` which is the base dependency for any engine but use JUnit 4 if `junit:junit` is present unless `org.junit.vintage:junit-vintage-engine` is also present. For TestNG it's similar because groovy-all also includes a dependency on it and we still should default to JUnit 4 if it's also present.

Thus, I don't think it's going to get much simpler than this (using coordinates instead of file names, of course):

```java
if (artifacts.contains("junit-platform-engine")
        && (!artifacts.contains("junit") || artifacts.contains("junit-vintage-engine"))) {
    // Use JUnit Platform
} else if (artifacts.contains("testng") && !artifacts.contains("junit")) {
    // Use TestNG
} else {
    // Use JUnit
}
```

> Maybe we'd also want to make spockframework auto-detect as JUnit.

I wouldn't do that because Spock 2.0 will use the JUnit Platform.